### PR TITLE
Allow installing newer containerd runc on cos

### PR DESCRIPTION
**Pull request by dims**
Currently COS images have embedded versions of containerd and runc which we use as-is. Unfortunately when we fix bugs in containerd we do not have a way to update the CI jobs to pick them up. Here we add the same mechanism we have for containerd on ubuntu images, but for COS images. 

Example: [containerd v1.7.14](https://github.com/containerd/containerd/releases/tag/v1.7.14) has a fix we need:
https://github.com/containerd/containerd/pull/9905

Which fixes the issue we see here:
https://testgrid.k8s.io/google-gce#gci-gce-alpha-enabled-default&width=20

This set of options `KUBE_COS_INSTALL_CONTAINERD_VERSION` and `KUBE_COS_INSTALL_RUNC_VERSION` allows us to inject the newer versions to ensure we actually fixed the problem without waiting for COS images with those specific versions of containerd/runc to be available.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
🤔  ... @BenTheElder @bobbypage don't we have a better solution to this problem?

**Comment by dims** [comment]
> 🤔 ... @BenTheElder @bobbypage don't we have a better solution to this problem?

@aojea this is the exact same mechanism used in containerd jobs to test containerd master with COS images :) i just copied the idea from there.

**Comment by dims** [comment]
( if there's another way to do this we can certainly switch over, for now, this works! )

**Review by bobbypage — COMMENTED** [review]
Thanks for adding this! A few questions...

File: cluster/gce/gci/configure.sh [hunk 1]
```diff
@@ -542,6 +542,48 @@ function install-containerd-ubuntu {
...
+      --location \
+      "https://github.com/opencontainers/runc/releases/download/${COS_INSTALL_RUNC_VERSION}/runc.${HOST_ARCH}" --output /home/containerd/bin/runc \
+    && chmod 755 /home/containerd/bin/runc
+  fi
+  cp /usr/lib/systemd/system/containerd.service /etc/systemd/system/containerd.service
+
+  # fix the path of the new containerd binary
+  sed -i 's|ExecStart=.*|ExecStart=/home/containerd/bin/containerd|' /etc/systemd/system/containerd.service
+  # ensure runc gets picked up from the correct location
+  sed -i "/\[Service\]/a Environment=PATH=/home/containerd/bin:$PATH" /etc/systemd/system/containerd.service
```
[the remark this answers is not in the document]
Comment (dims) [reply cluster/gce/gci/configure.sh:581]: I didn't want to touch the containerd.toml here so updated the PATH. (i can verify again, but if it does not work, i can switch to what we do in the other repo)

File: cluster/gce/gci/configure.sh [hunk 2]
```diff
@@ -542,6 +542,48 @@ function install-containerd-ubuntu {
...
+  systemctl stop containerd
+  mkdir -p /home/containerd/
+  mount --bind /home/containerd /home/containerd
+  mount -o remount,exec /home/containerd
+  if [[ -n "${COS_INSTALL_CONTAINERD_VERSION:-}" ]]; then
+    # containerd versions have slightly different url(s), so try both
+    # shellcheck disable=SC2086
+    ( curl ${CURL_FLAGS} \
+        --location \
+        "https://github.com/containerd/containerd/releases/download/${COS_INSTALL_CONTAINERD_VERSION}/containerd-${COS_INSTALL_CONTAINERD_VERSION:1}-${HOST_PLATFORM}-${HOST_ARCH}.tar.gz" \
```
[the remark this answers is not in the document]
Comment (dims) [reply cluster/gce/gci/configure.sh:563]: compare with the the ubuntu, it's the same basically we drop the `v` in the name of the file part of the url `containerd-1.7.14-linux-amd64.tar.gz` 

```
https://github.com/containerd/containerd/releases/download/v1.7.14/containerd-1.7.14-linux-amd64.tar.gz
```

File: cluster/gce/gci/configure.sh [hunk 3]
```diff
@@ -542,6 +542,48 @@ function install-containerd-ubuntu {
...
+    curl ${CURL_FLAGS} \
+      --location \
+      "https://github.com/opencontainers/runc/releases/download/${COS_INSTALL_RUNC_VERSION}/runc.${HOST_ARCH}" --output /home/containerd/bin/runc \
+    && chmod 755 /home/containerd/bin/runc
+  fi
+  cp /usr/lib/systemd/system/containerd.service /etc/systemd/system/containerd.service
+
+  # fix the path of the new containerd binary
+  sed -i 's|ExecStart=.*|ExecStart=/home/containerd/bin/containerd|' /etc/systemd/system/containerd.service
+  # ensure runc gets picked up from the correct location
```
[the remark this answers is not in the document]
Comment (dims) [reply cluster/gce/gci/configure.sh:580]: correct.

[the remark this answers is not in the document]
Comment (dims) [reply cluster/gce/gci/configure.sh]: Done @bobbypage !

**Comment by dims** [comment]
@bobbypage addressed all the comments. should be ready now please peek!

File: cluster/gce/gci/configure.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (bobbypage) [reply cluster/gce/gci/configure.sh]: Gotcha, in theory this should work (containerd spawns the shim, which then spawns runc, so it makes sense that adjusting path of containerd should result in the shim picking up the custom runc). Would be good to test it to verify this works in practice though :)

**Comment by bobbypage** [comment]
LGTM, I think this makes sense if we need to pick up a newer containerd/runc before COS image is released. Small comment about the PATH change to pick up custom runc, but looks good!

/lgtm

**Pull request merged**